includes/classes/SiteAutomation/Auth.php

```diff
@@ -0,0 +1,132 @@
+<?php
+/**
+ * Request and manage access_token for API calls.
+ *
+ * @package SophiWP
+ */
+
+namespace SophiWP\SiteAutomation;
+
+use function SophiWP\Settings\get_sophi_settings;
+
+/**
+ * Class: Auth
+ */
+class Auth {
+	/**
+	 * Class constructor.
+	 */
+	public function __construct() {
+		$this->environment = get_sophi_settings( 'environment' );
+	}
+
+	/**
+	 * Get cached access_token.
+	 *
+	 * @return string|\WP_Error
+	 */
+	public function get_access_token() {
+		$access_token = get_transient( 'sophi_site_automation_access_token' );
+
+		if ( $access_token ) {
+			return $access_token;
+		}
+
+		return $this->refresh_access_token();
+	}
+
+	/**
+	 * Refresh the access_token and save it to the database.
+	 *
+	 * @return string|\WP_Error
+	 */
+	public function refresh_access_token() {
+		$response = $this->request_access_token();
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		set_transient( 'sophi_site_automation_access_token', $response['access_token'], $response['expires_in'] );
```

Comment:
Yes we can do that, should we use `md5` while saving the token in the database?

https://github.com/globeandmail/sophi-for-wordpress/blob/98299cf5c0b7657a699d7a11586184fcef74d52c/includes/classes/SiteAutomation/Auth.php#L50

And we then have to decrypt here:

https://github.com/globeandmail/sophi-for-wordpress/blob/98299cf5c0b7657a699d7a11586184fcef74d52c/includes/classes/SiteAutomation/Auth.php#L29

Reply:
So `md5` is used to hash something (like a password) but is not intended to be used when you need to be able to decrypt a value (which is what we'll need to do in this case).

I'm not an expert when it comes to this and I know there are multiple tools you can use to handle the decryption and multiple ways to deal with the storing of the encryption key. I researched this topic for something else we were working on and I found https://felix-arntz.me/blog/storing-confidential-data-in-wordpress/ to be super helpful here. I'd recommend taking a look at that for a possible approach we could take here.

Reply:
@peterwilsoncc tagging you here in case you've got any advice on potential approach here (unless Felix's notes best apply)?

Reply:
Felix's advice seems pretty good to me.

For the key, you can probably use `wp_salt( 'logged_in' )` rather than accessing the php constant direct. This will ensure that you don't end up encrypting with the value `put your unique phrase here`. That will negate to need for a fallback too.

Reply:
Oh, something else to keep in mind is that transients are not guaranteed to stick around for the length of time specified in the second parameter of `set_transient()`. In theory they can be deleted immediately (there are plugins that disable them altogether) but more likely they drop out of the cache earlier than expected on sites using a persistent object cache.

tl;dr: if this must stick around then it's better to store as an option `[key => $key, expires => $timestamp ]`. Transients are fine if the time to live is intended as a suggestion only.

Reply:
Thank you for the details every one. Implemented in: https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/e911f6a87029c6389d38d1607e8db771558ecda1

@peterwilsoncc even if the transient is deleted before the actual expiry, we create another one when not found. Storing it as an array in the option key would require: getting the array > check for expiry -> use the token, that would take a bit more time IMO.